lib/web/gziphandler.go

```diff
@@ -0,0 +1,47 @@
+package web
+
+import (
+	"compress/gzip"
+	"io"
+	"net/http"
+	"strings"
+)
+
+type gzipResponseWriter struct {
+	io.Writer
+	http.ResponseWriter
+}
+
+// Write uses the Writer part of gzipResponseWriter to write the output.
+func (w gzipResponseWriter) Write(b []byte) (int, error) {
+	_, haveType := w.Header()["Content-Type"]
+	// Explicitly set Content-Type if it has not been set previously.
+	if !haveType {
+		// If no content type, apply sniffing algorithm to un-gzipped body.
+		w.Header().Set("Content-Type", http.DetectContentType(b))
+	}
+	return w.Writer.Write(b)
+}
+
+// MakeGzipHandler adds support for gzip compression for given handler.
+func MakeGzipHandler(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Check if the client can accept the gzip encoding and that this is not an image asset.
+		if !strings.Contains(r.Header.Get("Accept-Encoding"), "gzip") || isCompressedImageRequest(r) {
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		// Set the HTTP header indicating encoding.
+		w.Header().Set("Content-Encoding", "gzip")
+		gz := gzip.NewWriter(w)
+		defer gz.Close()
```

Comment:
Instead of creating a new writer for every request can you reuse the writer? You can just cut/paste the code from below.

```
// writerPool is a sync.Pool for shared gzip writers.
// each gzip writer allocates a lot of memory
// so it makes sense to reset the writer and reuse the
// internal buffers to avoid too many objects on the heap
```

https://github.com/gravitational/teleport/blob/master/lib/events/sessionlog.go#L615-L633